apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/postprocessor/AssignToPostProcessorFactory.java

```diff
@@ -108,33 +112,59 @@ public StackManipulation resolve(TypeDescription typeDescription, MethodDescript
         return new Advice.PostProcessor() {
             @Override
             public StackManipulation resolve(TypeDescription instrumentedType, MethodDescription instrumentedMethod, Assigner assigner, Advice.ArgumentHandler argumentHandler) {
-                final TypeDescription.Generic returnType = adviceMethod.getReturnType();
                 if (assignToReturn.index() != -1) {
-                    if (!returnType.represents(Object[].class)) {
-                        throw new IllegalStateException("Advice method has to return Object[] when setting an index");
-                    }
-                    return new StackManipulation.Compound(
-                        MethodVariableAccess.REFERENCE.loadFrom(exit ? argumentHandler.exit() : argumentHandler.enter()),
-                        IntegerConstant.forValue(assignToReturn.index()),
-                        ArrayAccess.REFERENCE.load(),
-                        assigner.assign(TypeDescription.Generic.OBJECT, instrumentedMethod.getReturnType(), Assigner.Typing.DYNAMIC),
-                        MethodVariableAccess.of(instrumentedMethod.getReturnType()).storeAt(argumentHandler.returned())
-                    );
+                    return objectArrayStackManipulation(adviceMethod, assigner, false, argumentHandler, exit, assignToReturn.index(), instrumentedMethod.getReturnType(), MethodVariableAccess.of(instrumentedMethod.getReturnType()).storeAt(argumentHandler.returned()));
```

Comment:
[minor] maybe reformat to have one parameter per line for methods calls that have plenty of parmeters/builders.